src/Features/CSharp/Portable/InlineTypeCheck/CSharpInlineTypeCheckAnalyzer.cs

```diff
@@ -0,0 +1,163 @@
+                ? (IdentifierNameSyntax)condition.Left
+                : (IdentifierNameSyntax)condition.Right;
+
+            // Now make sure the previous statement is "var a = ..."
+            var parentBlock = (BlockSyntax)ifStatement.Parent;
+            var ifIndex = parentBlock.Statements.IndexOf(ifStatement);
+            if (ifIndex == 0)
+            {
+                return;
+            }
+
+            var previousStatement = parentBlock.Statements[ifIndex - 1];
+            if (!previousStatement.IsKind(SyntaxKind.LocalDeclarationStatement))
+            {
+                return;
+            }
+
+            var localDeclarationStatement = (LocalDeclarationStatementSyntax)previousStatement;
+            var variableDeclaration = localDeclarationStatement.Declaration;
+
+            if (variableDeclaration.Variables.Count != 1)
+            {
+                return;
+            }
+
+            var declarator = variableDeclaration.Variables[0];
+            if (declarator.Initializer == null)
+            {
+                return;
+            }
+
+            if (!Equals(declarator.Identifier.ValueText, conditionName.Identifier.ValueText))
+            {
+                return;
+            }
+
+            // Make sure the initializer has the form "... = expr as Type;
+            var initializerValue = declarator.Initializer.Value;
+            if (!initializerValue.IsKind(SyntaxKind.AsExpression))
+            {
+                return;
+            }
+
+            // Looks good!
+            var additionalLocations = ImmutableArray.Create(
+                localDeclarationStatement.GetLocation(),
+                ifStatement.GetLocation(),
+                condition.GetLocation(),
+                initializerValue.GetLocation());
+
+            // Put a diagnostic with the appropriate severity on the declaration-statement itself.
+            syntaxContext.ReportDiagnostic(Diagnostic.Create(
+                CreateDescriptor(this.DescriptorId, severity),
+                localDeclarationStatement.GetLocation(),
+                additionalLocations));
+        }
+
+        private BinaryExpressionSyntax GetLeftmostCondition(ExpressionSyntax condition)
+        {
+            switch (condition.Kind())
+            {
+                case SyntaxKind.ParenthesizedExpression:
+                    return GetLeftmostCondition(((ParenthesizedExpressionSyntax)condition).Expression);
+                case SyntaxKind.ConditionalExpression:
+                    return GetLeftmostCondition(((ConditionalExpressionSyntax)condition).Condition);
+                case SyntaxKind.LogicalAndExpression:
+                case SyntaxKind.LogicalOrExpression:
+                    return GetLeftmostCondition(((BinaryExpressionSyntax)condition).Left);
+            }
+
+            return condition as BinaryExpressionSyntax;
+        }
+
+        private bool IsNullCheckExpression(ExpressionSyntax left, ExpressionSyntax right) =>
```

Comment:
I would probably name these parameters differently since you call it with left-right and right-left. Maybe `expression1` and `expression2`?

Also, isn't your style to put the arrow to the left of the expression for expression bodied members like this?